lib/checkunusedvar.cpp

```diff
@@ -1586,3 +1589,67 @@ bool CheckUnusedVar::isEmptyType(const Type* type)
     emptyType=false;   // unknown types are assumed to be nonempty
     return emptyType;
 }
+
+bool CheckUnusedVar::isFunctionWithoutSideEffects(const Function& func) {
+    // no body to analyze
+    if (!func.hasBody()) {
+        return false;
+    }
+
+    bool sideEffectReturnFound = false;
+    for (Token* bodyToken = func.functionScope->bodyStart->next(); bodyToken != func.functionScope->bodyEnd; 
+        bodyToken = bodyToken->next())
+    {
+        const Variable* bodyVariable = bodyToken->variable();
+        if (bodyVariable) {
+            // check variable for side-effects
+            if (!isVariableWithoutSideEffects(*bodyVariable)) {
+                return false;
+            }
+            // check if global variable is changed
+            if (bodyVariable->isGlobal()) {
```

Comment:
I have the feeling we need to be more careful about global variables

    buf[12] = 3;

Reply:
I've added a test case for this one.